File: 2025/zig/src/day10.zig
const std = @import("std");
const adlib = @import("adlib.zig");

const ButtonState = packed struct(u64) {
    state: u10 = 0,
    presses: u54 = 0,
};

fn partOne(input: *std.Io.Reader) !u64 {
    var total_min_presses: u64 = 0;
    var processed = std.AutoHashMap(u10, void).init(adlib.allocator);
    var queue = try adlib.Queue(ButtonState).initCapacity(adlib.allocator, 300);
    var linenr: u32 = 0;
    while (try input.takeDelimiter('\n')) |line| {
        linenr += 1;
        const close_bracket = std.mem.indexOfScalarPos(u8, line, 1, ']').?;
        var target: u10 = 0;
        for (1..close_bracket) |i| {
            if (line[i] == '#') {
                const bp = @as(u4, 10) - @as(u4, @truncate(i));
                target |= @as(u10, 1) << bp;
            }
        }

        var button_buf: [32]u10 = @splat(0);
        var button_count: u5 = 0;
        var par_start_idx = std.mem.indexOfScalarPos(u8, line, close_bracket + 1, '(');
        while (par_start_idx) |par_start| {
            const par_end = std.mem.indexOfScalarPos(u8, line, par_start + 1, ')').?;
            var nums = std.mem.tokenizeScalar(u8, line[par_start + 1 .. par_end], ',');
            var button: u10 = 0;
            while (nums.next()) |num_str| {
                const num = std.fmt.parseInt(u4, num_str, 10) catch unreachable;
                button |= @as(u10, 1) << (@as(u4, 9) - num);
            }
            button_buf[button_count] = button;
            button_count += 1;
            par_start_idx = std.mem.indexOfScalarPos(u8, line, par_end + 1, '(');
        }
        const buttons = button_buf[0..button_count];

        var min_presses: u64 = std.math.maxInt(u64);
        var max_queue_len: usize = 0;
        queue.push(.{});
        try processed.put(0, {});

        while (!queue.isEmpty()) {
            const item = queue.pop().?;
            if (item.state == target) {
                min_presses = @min(item.presses, min_presses);
                continue;
            }
            if (item.presses >= min_presses) continue;
            for (buttons) |button| {
                const new_state = item.state ^ button;
                if (!processed.contains(new_state)) {
                    try processed.put(item.state, {});
                    queue.push(.{ .state = new_state, .presses = item.presses + 1 });
                }
            }

            max_queue_len = @max(max_queue_len, queue.len());
        }

        if (min_presses != std.math.maxInt(u64))
            total_min_presses += min_presses;
        queue.clear();
        processed.clearRetainingCapacity();
    }

    queue.deinit();
    processed.deinit();
    return total_min_presses;
}

fn partTwo(input: *std.Io.Reader) !u32 {
    _ = input;
    return 0;
}

pub fn main() !void {
    var buf: [4096]u8 = undefined;
    var in_file = try adlib.inputFile("10");
    var reader = in_file.reader(&buf);
    const res_1 = try partOne(&reader.interface);
    try reader.seekTo(0);
    const res_2 = try partTwo(&reader.interface);
    std.debug.print("part one: {d}\npart two: {d}\n", .{ res_1, res_2 });
}

File: 2025/rust/src/bin/day10.rs
use std::collections::{HashSet, VecDeque};

#[derive(Debug)]
struct ButtonState {
    state: u16,
    presses: u32,
}

fn part_one(input: &String) -> u64 {
    let mut total_min_presses = 0;
    let mut processed: HashSet<u16> = HashSet::new();
    let mut queue = VecDeque::new();

    for line in input.lines() {
        let close_bracket = line.find(']').unwrap();
        let target_len = close_bracket - 1;
        let mut target_state = 0;
        let line_bytes = line.as_bytes();
        for i in 1..close_bracket {
            if line_bytes[i] == b'#' {
                target_state |= 1 << (target_len - i);
            }
        }
        let mut buttons: Vec<u16> = vec![];

        let brack_end = line.find(']').unwrap();
        let brace_start = line.find('{').unwrap();
        for button_str in line[brack_end + 1..brace_start].split(' ') {
            if button_str.len() == 0 {
                continue;
            }
            let mut button = 0;
            for digit_str in button_str.matches(char::is_numeric) {
                let num: usize = digit_str.parse().unwrap();
                button |= 1 << (target_len - 1 - num);
            }
            buttons.push(button);
        }

        let mut min_presses = u64::MAX;
        queue.push_back(ButtonState {
            state: 0,
            presses: 0,
        });
        processed.insert(0);

        while !queue.is_empty() {
            let next = queue.pop_front().unwrap();
            if next.state == target_state {
                min_presses = min_presses.min(next.presses as u64);
                continue;
            }
            if (next.presses as u64) >= min_presses {
                continue;
            }

            for button in buttons.iter() {
                let new_state = next.state ^ button;
                if !processed.contains(&new_state) {
                    processed.insert(new_state);
                    queue.push_back(ButtonState {
                        state: new_state,
                        presses: next.presses + 1,
                    });
                }
            }
        }

        if min_presses != u64::MAX {
            total_min_presses += min_presses;
        }
        queue.clear();
        processed.clear();
    }

    return total_min_presses;
}

pub fn main() -> Result<(), std::io::Error> {
    let input = std::fs::read_to_string("input/10")?;
    let res_1 = part_one(&input);
    println!("part_one: {}", res_1);
    Ok(())
}
